LFS-333: Integrate NCR

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-333)

**Test branch**: `LFS-333-test` 

**Before testing**: **Note that this will not work within our test deployment VMs, since this requires you to run the entire Docker pipeline with NCR enabled (see the base `Readme.md` for instructions on how to accomplish this)**. Steps (copied from `Readme.md`):
1) Clone [https://github.com/ccmbioinfo/NeuralCR](https://github.com/ccmbioinfo/NeuralCR)
2) Checkout the **develop** branch.
3) `docker build -t ccmsk/neuralcr .`
4) Download the pretrained ncr model parameters (Be in your Downloads folder for this)
```
wget https://github.com/ccmbioinfo/NeuralCR/releases/download/1.0/ncr_model_params.tar.gz
tar -xvf ncr_model_params.tar.gz
```
5) Build LFS as normal: (Be in `lfs/` for this) `mvn clean install`
6) Now we use docker-compose to build things for us:
```
cd compose-cluster
python3 generate_compose_yaml.py --shards 2 --replicas 1 --enable_ncr
docker-compose build
docker-compose up -d
```
7) **When stopping the instance, follow all the following**:
```
docker-compose down #Stop all containers
docker-compose rm #Remove all stopped containers
docker volume prune -f #Remove all stored data
./cleanup.sh #Remove the cluster configuration files
```

**To test**: From Vocabularies on the sidebar, install HPO. Create a Demographics form, then scroll down to Co-morbidities. Add a note, then click away from the note. A new note will appear on the right side with the NCR's best interpretation of your input. Clicking on individual terms will add them to the answer.

**Additional notes**: The `docker-compose up -d` command seems to be very stochastic, working about half the time. Also note that installing HPO on MongoDB takes a long while (30+ minutes).

OrdiNeu committed:
LFS-333: Integrate NCR
Add a new component, NCRNote, to store notes that can be parsed through NCR
Clicking on a vocabulary term on the parsed output adds it to the selection

OrdiNeu committed:
LFS-333: Integrate NCR
Code cleanup

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/NCRNote.jsx:
```diff
@@ -0,0 +1,184 @@
+      >
+      {isLoading && <CircularProgress className={classes.NCRLoadingIndicator} />}
+      {parsedText}
+      {error && <div>{error}</div>}
```
The error message should be formatted (red text, the usual error typography, etc).

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Answer.jsx:
```diff
@@ -64,11 +65,21 @@ function Answer (props) {
-          answerPath={answerPath}
-          onChangeNote={onChangeNote}
-          />}
+        (sourceVocabulary ?
```
I'm not a fan of this approach. Two issues:
* NCR is not truly an optional component if we hardcore the corresponding ui
* `Answer` should remain generic and not deal with cases that are specific to a certain answer type.

I don't know yet how to approach this using React (innpure JS decoupling is easy but won't work as nicely here. To be discussed.

OrdiNeu committed:
LFS-333: Integrate NCR
Refactor NCRNote out of Answer and into vocabSelector

OrdiNeu committed:
LFS-333: Integrate NCR
Move the error message from NCRNotes into an error Typography

OrdiNeu committed:
LFS-333: Integrate NCR
Bugfix where only the first suggestion was selectable.

OrdiNeu committed:
LFS-333: Integrate NCR
Properly handle state update requests in select.
Gracefully handle 404s when a vocabulary question's vocabulary is
missing.

sashaandjic reviewed (approved):
Tested.

IntegralProgrammer commented on modules/data-entry/src/main/frontend/src/questionnaire/NCRNote.jsx:
```diff
@@ -0,0 +1,186 @@
+    // Access the NCR URL
+    let url = new URL(NCRURL);
+    url.searchParams.set("text", cachedText);
+    // NCR calls HP HPO, so we do a quick conversion here
```
This should not be merged - instead, there should only be a call to `url.searchParams.set("model", vocabulary)` and within *ccmbioinfo/NeuralCR*, inside `app.py`, `NCR_MODELS['HPO']` should be renamed to `NCR_MODELS['HP']`.

OrdiNeu replied:
This is now in PR at https://github.com/ccmbioinfo/NeuralCR/pull/14

IntegralProgrammer commented on modules/data-entry/src/main/frontend/src/questionnaire/NCRNote.jsx:
```diff
@@ -0,0 +1,186 @@
+    var firstMatch = matches[0];
+    // The front matter is the text before the first match begins
+    frontMatter = text.substring(0, firstMatch.start-offset);
+    var matchID = firstMatch["hp_id"]; // TODO: Handle more than just HPO
```
The `hp_id` key is not unique to the Human Phenotype Ontology (HPO) language model. For example, using the MONDO ontology will also give results with a `hp_id` key. In future updates to *NeuralCR*, `hp_id` should be replaced with something more meaningful, but for now, create a `const` variable named `ONTOLOGY_KEY` and set it equal to `hp_id`. Therefore, when *NeuralCR* is updated, we will simply need to update the value of the `ONTOLOGY_KEY` constant.

IntegralProgrammer commented on modules/data-entry/src/main/frontend/src/questionnaire/NCRNote.jsx:
```diff
@@ -0,0 +1,186 @@
+  // Handle the user clicking on a link which corresponds to a suggestion
+  let addSuggestion = (event) => {
+    event.preventDefault();
+    onAddSuggestion(firstMatch["hp_id"], firstMatch["names"][0])
```
Replace `"hp_id"` with `ONTOLOGY_KEY`

OrdiNeu committed:
LFS-333: Integrate NCR
Move the string "hp_id" to a const

IntegralProgrammer commented:
Another small bug - not detrimental to the usability of LFS but should be fixed if time permits:

1. Build and start the `LFS-333-test` branch
2. Create a new *Demographics* form
3. Under *co-morbidities (non cancerous)* add a note `The patient is near-sighted`.
4. Click outside of the text box
5. On the right, `near-sighted` should be highlighted in *blue* - click it.
6. *Myopia* should be added to the list of co-morbidities
7. Click *Save*
8. Go to the *User dashboard*
9. From the *Demographics* card, re-open the form from the earlier steps
10. Click the *co-morbidities note* text box
11. Delete *Myopia* from the list of co-morbidities
12. On the right of the note box, click `near-sighted` highlighted in *blue*.
13. ***Myopia* is not added back to the *co-morbidities* list.**
14. Click inside the note box
15. Click outside the note box
16. On the right, click `near-sighted` highlighted in *blue*.
17. Now, *Myopia* is added back to the list of co-morbidities as we would expect it to.

OrdiNeu committed:
LFS-333: Integrate NCR
Code cleanup

OrdiNeu committed:
LFS-333: Integrate NCR
Fix a bug where sufficiently nested functions with access to
Select:onAddSelection would work off of initial Select state, allowing
duplicates to be added indefinitely.

OrdiNeu commented:
@IntegralProgrammer 
> Another small bug - not detrimental to the usability of LFS but should be fixed if time permits: 

Turns out that the small bug was indicative of a larger bug, where NCRNote had access to closures that were out of date, causing strange logic (duplicates being added indefinitely, etc). I've fixed it as best as I could in `b70309a`

IntegralProgrammer commented on modules/data-entry/src/main/frontend/src/vocabSelector/select.jsx:
```diff
@@ -145,32 +145,35 @@ function VocabularySelector(props) {
-    }
+    // Prevent closures from mucking up logic by placing everything in an updater function
+    setListChildren((oldChildren) => {
+      console.log(id);
```
Are these `console.log` statements still necessary, or were they only there for debugging?

IntegralProgrammer reviewed (changes requested):
Re-tested and the bug seems to be gone. Please have a look at the `console.log` statements, but other than that, I'll approve.

OrdiNeu committed:
[misc] Remove debugging code.
